quill-jdbc-zio/src/main/scala/io/getquill/context/qzio/ZioJdbcContext.scala

```diff
@@ -238,19 +239,19 @@ abstract class ZioJdbcContext[+Dialect <: SqlIdiom, +Naming <: NamingStrategy]
       case Some(connection) =>
         blocking(qlio.provideEnvironment(ZEnvironment(connection)))
       case None =>
-        blocking(qlio.provideLayer(Quill.Connection.acquireScoped))
+        blocking(qlio.provideLayer(Quill.Connection.acquire))
     }
 
   private def onConnectionStream[T](
     qstream: ZStream[Connection, SQLException, T]
   ): ZStream[DataSource, SQLException, T] =
-    streamBlocker *> ZStream.fromZIO(currentConnection.get).flatMap {
-      case Some(connection) =>
-        qstream.provideEnvironment(ZEnvironment(connection))
-      case None =>
-        (for {
-          env <- ZStream.scoped(Quill.Connection.acquireScoped.build)
-          r   <- qstream.provideEnvironment(env)
-        } yield (r)).refineToOrDie[SQLException]
+    ZStream.unwrapScoped[DataSource] {
+      for {
+        _               <- streamBlocker
+        maybeConnection <- currentConnection.get
+      } yield maybeConnection match {
+        case Some(connection) => qstream.provideEnvironment(ZEnvironment(connection))
+        case None             => qstream.provideLayer(Quill.Connection.acquire).refineToOrDie[SQLException]
+      }
```

Comment:
We need to write the code this way because of https://github.com/zio/zio-quill/pull/2928#discussion_r1366822824

If we want the stream to be executed on the blocking thread pool (which is why there's this `streamBlocker` call) all this code must be wrapped in the same Scope, which will control the executor shift back.